### PR TITLE
Donation vault

### PR DESCRIPTION
**Pull request by lukasz-zimnoch**
Closes: #237 

BTC donations are needed to re-balance the system in some rare cases when the BTC:tBTC peg can be breached. This change introduces a new `DonationVault` that is meant to handle BTC donations. That vault is plugged-in into the deposit flow. An external actor willing to make a BTC donation just makes a regular deposit targeting the `DonationVault`. Once that deposit is swept by the target wallet, the depositor's tBTC balance is not increased in the Bank, and the deposit is routed to the `DonationVault` which ignores it as well. In effect, deposited BTC remains swept by the wallet but no tBTC-related side effects take place.

### CODE REVIEW

**Comment by pdyraga** [comment]
👀

File: solidity/contracts/vault/DonationVault.sol [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+pragma solidity ^0.8.9;
+
+import "./IVault.sol";
+import "../bank/Bank.sol";
+
+/// @title BTC donation vault
+/// @notice Vault that allows making BTC donations to the system. Upon deposit,
+///         this vault does not increase depositors' tBTC balances and always
+///         decreases its own tBTC balance in the same transaction.
```
Comment (pdyraga) [inline solidity/contracts/vault/DonationVault.sol:24]:
```suggestion
/// @notice Vault that allows making BTC donations to the system. Upon deposit,
///         this vault does not increase depositor's balances and always
///         decreases its own balance in the same transaction.
```

TBTC is only relevant for `TBTCVault`. All other system components operate on a balance in the Bank.

File: solidity/contracts/vault/DonationVault.sol [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+///
+///         BEWARE: ALL BTC DEPOSITS TARGETING THIS VAULT ARE NOT REDEEMABLE AND
+///         THERE IS NO WAY TO RESTORE THE TBTC BALANCE! DO NOT DEPOSIT AGAINST
```
Comment (pdyraga) [inline solidity/contracts/vault/DonationVault.sol:27]:
```suggestion
///         THERE IS NO WAY TO RESTORE THE BALANCE! DO NOT DEPOSIT AGAINST
```

File: solidity/contracts/vault/DonationVault.sol [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        external
+        override
+        onlyBank
+    {
+        revert("Donation vault cannot receive balance approval");
+    }
+
+    /// @notice Ignores the deposited amounts and does not increase depositors'
+    ///         individual tBTC balances. Decreases its own tBTC balance
+    ///         in the Bank by the total deposited amount.
```
Comment (pdyraga) [inline solidity/contracts/vault/DonationVault.sol:59]:
```suggestion
    /// @notice Ignores the deposited amounts and does not increase depositors'
    ///         individual balances. The vault decreases its own balance
    ///         in the Bank by the total deposited amount.
```

File: solidity/contracts/vault/DonationVault.sol [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        bank = _bank;
+    }
+
+    /// @notice Function not supported by the `DonationVault`. Always reverts.
+    function receiveBalanceApproval(address, uint256)
+        external
+        override
+        onlyBank
+    {
+        revert("Donation vault cannot receive balance approval");
```
Comment (pdyraga) [inline solidity/contracts/vault/DonationVault.sol:54]: Why not to allow someone make a donation from their Bank balance? 

I think we should allow it by `Bank.approveBalanceAndCall` as well as implement `DonationVault.donate()` for already approved balances.

Comment (lukasz-zimnoch) [reply solidity/contracts/vault/DonationVault.sol]: Yep, good idea! Here it is: https://github.com/keep-network/tbtc-v2/pull/269/commits/4f2b9d0bcc28ff55a08e12881d2c28331b3000e0

File: solidity/contracts/vault/DonationVault.sol (hunk 3 above)
Comment (lukasz-zimnoch) [reply solidity/contracts/vault/DonationVault.sol]: See 2efa4ced

File: solidity/contracts/vault/DonationVault.sol (hunk 2 above)
Comment (lukasz-zimnoch) [reply solidity/contracts/vault/DonationVault.sol]: See 2efa4ced

File: solidity/contracts/vault/DonationVault.sol (hunk 1 above)
Comment (lukasz-zimnoch) [reply solidity/contracts/vault/DonationVault.sol]: See 2efa4ced

**Comment by lukasz-zimnoch** [comment]
Ready for another shot @pdyraga !

File: solidity/contracts/vault/DonationVault.sol [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+/// @title BTC donation vault
+/// @notice Vault that allows making BTC donations to the system. Upon deposit,
+///         this vault does not increase depositors' balances and always
+///         decreases its own balance in the same transaction. The vault also
+///         allows making donations using existing Bank balances.
+///
+///         BEWARE: ALL BTC DEPOSITS TARGETING THIS VAULT ARE NOT REDEEMABLE
+///         AND THERE IS NO WAY TO RESTORE THE DONATED BALANCE! DO NOT DEPOSIT
+///         AGAINST THIS VAULT OR DONATE YOUR EXISTING BANK BALANCE UNLESS YOU
+///         REALLY KNOW WHAT YOU ARE DOING!
```
Comment (pdyraga) [inline solidity/contracts/vault/DonationVault.sol:30]: Let's simplify it a bit. I was really confused with *or* and *unless* when I was reading it for the first time :)

```suggestion
///         BEWARE: ALL BTC DEPOSITS TARGETING THIS VAULT ARE NOT REDEEMABLE
///         AND THERE IS NO WAY TO RESTORE THE DONATED BALANCE.
///         USE THIS VAULT ONLY WHEN YOU REALLY KNOW WHAT YOU ARE DOING!
```

File: solidity/contracts/vault/DonationVault.sol [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+contract DonationVault is IVault {
+    Bank public bank;
+
+    event DonationReceived(address donator, uint256 donatedAmount);
```
Comment (pdyraga) [inline solidity/contracts/vault/DonationVault.sol:34]: Can we replace `donator` with `donor`? The second option is a way most common word.

File: solidity/test/vault/DonationVault.test.ts [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+          .connect(account1)
+          .approveBalanceAndCall(vault.address, 1000)
+      })
+
+      after(async () => {
+        await restoreSnapshot()
+      })
+
+      it("should decrease donator's balance", async () => {
+        expect(await bank.balanceOf(account1.address)).to.be.equal(0)
```
Comment (pdyraga) [inline solidity/test/vault/DonationVault.test.ts:198]: Can we `increaseBalance` for a bit more than 1000 to confirm just the amount from `approveBalanceAndCall` is donated?

File: solidity/test/vault/DonationVault.test.ts [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+    context("when called with correct parameters", () => {
+      let tx: ContractTransaction
+
+      before(async () => {
+        await createSnapshot()
+
+        await bank.connect(bridge).increaseBalance(account1.address, 1000)
+        await bank
+          .connect(account1)
+          .increaseBalanceAllowance(vault.address, 1000)
```
Comment (pdyraga) [inline solidity/test/vault/DonationVault.test.ts:122]: Can we `increaseBalance` and `increaseBalanceAllowance` for a bit more than `1000` to confirm the amount passed as `donate` parameter is taken into account?

Comment (lukasz-zimnoch) [reply solidity/test/vault/DonationVault.test.ts]: See 52fc175

File: solidity/contracts/vault/DonationVault.sol (hunk 6 above)
Comment (lukasz-zimnoch) [reply solidity/contracts/vault/DonationVault.sol]: See https://github.com/keep-network/tbtc-v2/pull/269/commits/63eeb9baeb7f699785b9c3111102a0324ef54513

File: solidity/contracts/vault/DonationVault.sol (hunk 5 above)
Comment (lukasz-zimnoch) [reply solidity/contracts/vault/DonationVault.sol]: See https://github.com/keep-network/tbtc-v2/pull/269/commits/1624c3a7b29da55b57d9701840a0cd230c2ee3b8

**Pull request merged**